File: app/src/main/java/com/snazhmudinov/movies/adapters/MoviesAdapter.java
package com.snazhmudinov.movies.adapters;

import android.content.Context;
import android.net.Uri;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.facebook.drawee.view.SimpleDraweeView;
import com.snazhmudinov.movies.R;
import com.snazhmudinov.movies.constans.Constants;
import com.snazhmudinov.movies.interfaces.MovieInterface;
import com.snazhmudinov.movies.models.Movie;
import java.util.List;
import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by snazhmudinov on 5/28/17.
 */

public class MoviesAdapter extends RecyclerView.Adapter<MoviesAdapter.MovieHolder> {

    private List<Movie> moviesList;
    private Context mContext;
    private boolean isLocalImage = false;
    public MovieInterface movieInterface;

    public static final int GRID_MODE = 0;
    public static final int LIST_MODE = 1;
    private int mMode = 0;
    private int mCurrentSelection = 0;

    public MoviesAdapter(List<Movie> movies, Context context) {
        moviesList = movies;
        mContext = context;
    }

    @Override
    public MovieHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(mContext);
        View view = null;
        switch(viewType) {
            case GRID_MODE:
                view = inflater.inflate(R.layout.movie_grid_item, parent, false);
                break;

            case LIST_MODE:
                view = inflater.inflate(R.layout.movie_list_item, parent, false);
                break;
        }


        return new MovieHolder(view);
    }

    @Override
    public void onBindViewHolder(MovieHolder holder, int position) {
        Movie currentMovie = moviesList.get(position);
        Uri uri = isLocalImage ? Uri.parse(currentMovie.getPosterPath()) :
                Uri.parse(Constants.POSTER_BASE_URL + currentMovie.getPosterPath());

        holder.mPosterView.setImageURI(uri);

        if (mMode == LIST_MODE) {
            final TextView tv = (TextView) holder.itemView.findViewById(R.id.movie_title);
            tv.setText(moviesList.get(position).getTitle());
            holder.itemView.setSelected(position == mCurrentSelection);
        }
    }

    @Override
    public int getItemViewType(int position) {
        return mMode;
    }

    @Override
    public int getItemCount() {
        return moviesList.size();
    }

    class MovieHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.poster)
        SimpleDraweeView mPosterView;

        MovieHolder(final View itemView) {
            super(itemView);

            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Movie movie = moviesList.get(getAdapterPosition());
                    if (movieInterface != null) {
                        mCurrentSelection = moviesList.indexOf(movie);
                        movieInterface.onMovieSelected(movie, isLocalImage);
                        notifyDataSetChanged();
                    }
                }
            });
        }
    }

    public void setLocalImage(boolean value) {
        isLocalImage = value;
    }

    public void setMode(int mode) {
        mMode = mode;
    }

    public void setSelectionIndex(int index) {
        mCurrentSelection = index;
    }
}

File: app/src/main/java/com/snazhmudinov/movies/adapters/MoviesAdapter.kt
package com.snazhmudinov.movies.adapters

import android.content.Context
import android.net.Uri
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.snazhmudinov.movies.R
import com.snazhmudinov.movies.constans.Constants
import com.snazhmudinov.movies.models.Movie
import com.snazhmudinov.movies.modules.GlideApp
import kotlinx.android.synthetic.main.movie_rv_item_list.view.*
import java.text.DateFormatSymbols

/**
 * Created by snazhmudinov on 5/28/17.
 */

class MoviesAdapter(private val moviesList: List<Movie>, private val mContext: Context, private val isTablet: Boolean) : RecyclerView.Adapter<MoviesAdapter.MovieHolder>() {

    var movieInterface: MovieInterface? = null
    var indexOfSelectedMovie: Int = 0
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    interface MovieInterface {
        fun onMovieSelected(movie: Movie)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieHolder {
        val inflater = LayoutInflater.from(mContext)
        val view: View = if (isTablet) {
            inflater.inflate(R.layout.movie_rv_item_list, parent, false)
        } else {
            inflater.inflate(R.layout.movie_rv_item_tile, parent, false)
        }

        return MovieHolder(view)
    }

    override fun onBindViewHolder(holder: MovieHolder, position: Int) {
        val currentMovie = moviesList[position]
        val uri = if (currentMovie.savedFilePath == null) { Uri.parse(Constants.POSTER_BASE_URL + currentMovie.posterPath) }
                    else { Uri.parse(currentMovie.savedFilePath) }

        GlideApp.with(mContext).load(uri).into(holder.itemView.poster)
        holder.itemView.movie_title?.text = currentMovie.originalTitle
        holder.itemView.additional_info?.text = formatMovieDate(currentMovie)

        val backgroundColor = if (position == indexOfSelectedMovie) ContextCompat.getColor(mContext, R.color.gray_selector)
                              else ContextCompat.getColor(mContext, android.R.color.transparent)
        holder.itemView.setBackgroundColor(backgroundColor)
    }

    override fun getItemCount() = moviesList.size

    inner class MovieHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        init { itemView.setOnClickListener { openMovieDetails() } }

        private fun openMovieDetails() {
            val movie = moviesList[adapterPosition]
            indexOfSelectedMovie = adapterPosition
            movieInterface?.onMovieSelected(movie)
        }
    }

    private fun formatMovieDate(movie: Movie): String {
        val movieDate = movie.releaseDate
        val dateComponents = movieDate.split("-")

        val dfs = DateFormatSymbols()

        val day = dateComponents[2]
        val month = dfs.months[dateComponents[1].toInt() - 1]
        val year = dateComponents[0]

        return "$month $day, $year"
    }
}
